aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,158 @@
+
+:doc:`Learn more about controllers and actions <actions>`.
+
+Controller logic should be minimal and should not be focused on business logic or infrastructure concerns like data access. When testing controller logic, avoid testing the framework as much as possible. You should test how the controller *behaves* based on valid or invalid inputs, and what kind of response it provides based on the result of some business operation it performs.
+
+Typical controller responsibilities:
+	- Verify ``ModelState.IsValid``
+	- Return an error response if ``ModelState`` is invalid
+	- Retrieve a business entity from persistence
+	- Perform an action on the business entity
+	- Save the business entity to persistence
+	- Return an appropriate ``IActionResult``
+
+Unit Testing
+------------
+:doc:`Unit testing </testing/unit-testing>` involves testing a part of an app in isolation from its infrastructure and dependencies. When unit testing controller logic, only the contents of a single action should be tested, not the behavior of its dependencies or of the framework itself. Thus, when unit testing a controller's actions, avoid testing global or attribute-based :doc:`filters <filters>`, :doc:`routing </fundamentals/routing>`, or :doc:`model binding </mvc/models/model-binding>`, as these are all performed by the framework (testing these is an integration test responsibility).
+
+To demonstrate unit testing, review the following controller. It displays a list of brainstorming sessions and allows new brainstorming sessions to be created with a POST:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/HomeController.cs
+  :language: c#
+  :emphasize-lines: 11,15,20,40-41
+
+The controller is following the `explicit dependencies principle <http://deviq.com/explicit-dependencies-principle/>`_, expecting dependency injection to provide it with an instance of ``IBrainStormSessionRepository``. This makes it fairly easy to test using a mock object framework, like `Moq <https://www.nuget.org/packages/Moq/>`_. To test the first ``Index`` method, which has no looping or branching logic and only calls one method, we need only verify that a ``ViewResult`` is returned, with a ``ViewModel`` containing whatever was returned from the repository's ``List`` method.
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/HomeControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 16-17,23-25
+
+The second ``Index`` method (which accepts an ``HttpPost``) has slightly more logic, since it checks ``ModelState.IsValid``. We should test that, when ``ModelState.IsValid`` is ``false`` the action method returns a ``ViewResult`` with the appropriate data. Otherwise, we should confirm that the ``Add`` method is called on the repository and a ``RedirectToActionResult`` is returned with the correct arguments. Below are its unit tests:
```

Comment:
The ``HttpPost Index`` unit tests (shown below) should verify:

-  When ``ModelState.IsValid`` is ``false``, the action method returns a ``ViewResult`` with the appropriate data
-  When ``ModelState.IsValid`` is ``true``, the ``Add`` method is called on the repository and a ``RedirectToActionResult`` is returned with the correct arguments

delete "Below are its unit tests:"